oauth/iam/iam.go

```diff
@@ -0,0 +1,231 @@
+	ctxWithCancel, cancelFunc := context.WithCancel(context.Background())
+	iamsvc, err := iamv1.NewService(ctxWithCancel, opts...)
+	if err != nil {
+		cancelFunc()
+		return nil, fmt.Errorf("failed to create new IAM credentials service: %v", err)
+	}
+	return &IAMService{
+		ctx:        ctxWithCancel,
+		cancelFunc: cancelFunc,
+		svc:        iamsvc,
+	}, nil
+}
+
+func (s *IAMService) Close() {
+	s.cancelFunc()
+}
+
+// AccessTokenSource returns a new access token source. Service account email and scopes are required.
+func (s *IAMService) AccessTokenSource(saEmail string, scopes []string, refreshInterval time.Duration) (*AccessTokenSource, error) {
+	if saEmail == "" {
+		return nil, fmt.Errorf("service account is required to create access token source")
+	}
+	if len(scopes) == 0 {
+		return nil, fmt.Errorf("scopes are required to create access token source")
+	}
+	ats := &AccessTokenSource{
+		iamsvc: s.svc,
+		saName: fmt.Sprintf(serviceAccountNameFormat, saEmail),
+		scopes: scopes,
+	}
+
+	if refreshInterval == 0 {
+		refreshInterval = defaultRefreshInterval
+	}
+	go func() {
+		tick := time.NewTicker(refreshInterval)
+		for {
+			select {
+			case <-tick.C:
+				if err := ats.refresh(); err != nil {
+					log.Errorf("%v", err)
+				}
+			case <-s.ctx.Done():
+				tick.Stop()
+				return
+			}
+		}
+	}()
+
+	if err := ats.refresh(); err != nil {
+		log.Errorf("%v", err)
+	}
+	return ats, nil
+}
+
+// IdentityTokenSource returns a new ID token source. Service account email and audience are required.
+func (s *IAMService) IdentityTokenSource(saEmail, audience string, includeEmail bool, refreshInterval time.Duration) (*IdentityTokenSource, error) {
+	if saEmail == "" {
+		return nil, fmt.Errorf("service account is required to create id token source")
+	}
+	if audience == "" {
+		return nil, fmt.Errorf("audience is required to create id token source")
+	}
+	its := &IdentityTokenSource{
+		iamsvc:       s.svc,
+		saName:       fmt.Sprintf(serviceAccountNameFormat, saEmail),
+		audience:     audience,
+		includeEmail: includeEmail,
+	}
+
+	if refreshInterval == 0 {
+		refreshInterval = defaultRefreshInterval
+	}
+	go func() {
+		tick := time.NewTicker(refreshInterval)
+		for {
+			select {
+			case <-tick.C:
+				if err := its.refresh(); err != nil {
+					log.Errorf("%v", err)
+				}
+			case <-s.ctx.Done():
+				tick.Stop()
+				return
+			}
+		}
+	}()
+
+	if err := its.refresh(); err != nil {
+		log.Errorf("%v", err)
+	}
+	return its, nil
+}
+
+// Token returns the access token from the source.
+func (ats *AccessTokenSource) Token() (*oauth2.Token, error) {
+	if !ats.token.Valid() {
+		if err := ats.refresh(); err != nil {
+			return nil, err
+		}
+	}
+	ats.mu.Lock()
+	defer ats.mu.Unlock()
+	// Deep copy the object to avoid data race.
+	return &oauth2.Token{
+		AccessToken: ats.token.AccessToken,
+		Expiry:      ats.token.Expiry,
+	}, nil
+}
+
+func (ats *AccessTokenSource) refresh() error {
```

Comment:
I think you'll want to avoid a potential thundering herd issue in your refresh functions. See: https://github.com/apigee/apigee-remote-service-golib/blob/0bc2b615ebe95fb9b436bcdf765cac2e04ba66c2/auth/key/verify_api_key.go#L174.

Also, may need backoff() function to avoid rapid repeated failed callouts and cancel()  to allow quicker shutdown. See: https://github.com/apigee/apigee-remote-service-golib/blob/0bc2b615ebe95fb9b436bcdf765cac2e04ba66c2/auth/key/verify_api_key.go#L205.

Reply:
I believe the following is the case, is it not? Each incoming request calls EnvironmentSpecRequest.TargetAuth() which calls Token() which can call refresh() which will call GenerateAccessToken. You could have 1:1 such calls with incoming requests along this tree.

Reply:
Ah that makes sense! Added the herbBuster logic. PTAL.